src/layer/x86/lstm_x86.cpp

```diff
@@ -0,0 +1,933 @@
+                _sumG_1 = _mm256_fmadd_ps(_mm256_loadu_ps(weight_hc_G_1), h_cont, _sumG_1);
+                hidden_ptr_r += 8;
+                weight_hc_I_0 += 8;
+                weight_hc_F_0 += 8;
+                weight_hc_O_0 += 8;
+                weight_hc_G_0 += 8;
+                weight_hc_I_1 += 8;
+                weight_hc_F_1 += 8;
+                weight_hc_O_1 += 8;
+                weight_hc_G_1 += 8;
+            }
+            float sums[8];
+            _mm256_storeu_ps(sums, HorizontalSums(_sumI_0, _sumF_0, _sumO_0, _sumG_0, _sumI_1, _sumF_1, _sumO_1, _sumG_1));
+            sums[0] += bias_c_I[q];
+            sums[1] += bias_c_F[q];
+            sums[2] += bias_c_O[q];
+            sums[3] += bias_c_G[q];
+            sums[4] += bias_c_I[q + 1];
+            sums[5] += bias_c_F[q + 1];
+            sums[6] += bias_c_O[q + 1];
+            sums[7] += bias_c_G[q + 1];
+
+            for (; remain_size > 0; remain_size--)
+            {
+                float xi = *x;
+                sums[0] += *weight_xc_I_0 * xi;
+                sums[1] += *weight_xc_F_0 * xi;
+                sums[2] += *weight_xc_O_0 * xi;
+                sums[3] += *weight_xc_G_0 * xi;
+                sums[4] += *weight_xc_I_1 * xi;
+                sums[5] += *weight_xc_F_1 * xi;
+                sums[6] += *weight_xc_O_1 * xi;
+                sums[7] += *weight_xc_G_1 * xi;
+                x++;
+                weight_xc_I_0++;
+                weight_xc_F_0++;
+                weight_xc_O_0++;
+                weight_xc_G_0++;
+                weight_xc_I_1++;
+                weight_xc_F_1++;
+                weight_xc_O_1++;
+                weight_xc_G_1++;
+            }
+
+            for (; remain_num_output > 0; remain_num_output--)
+            {
+                float h_cont = *hidden_ptr_r;
+                sums[0] += *weight_hc_I_0 * h_cont;
+                sums[1] += *weight_hc_F_0 * h_cont;
+                sums[2] += *weight_hc_O_0 * h_cont;
+                sums[3] += *weight_hc_G_0 * h_cont;
+                sums[4] += *weight_hc_I_1 * h_cont;
+                sums[5] += *weight_hc_F_1 * h_cont;
+                sums[6] += *weight_hc_O_1 * h_cont;
+                sums[7] += *weight_hc_G_1 * h_cont;
+                hidden_ptr_r++;
+                weight_hc_I_0++;
+                weight_hc_F_0++;
+                weight_hc_O_0++;
+                weight_hc_G_0++;
+                weight_hc_I_1++;
+                weight_hc_F_1++;
+                weight_hc_O_1++;
+                weight_hc_G_1++;
+            }
+            gates_data_I[q] = sums[0];
+            gates_data_F[q] = sums[1];
+            gates_data_O[q] = sums[2];
+            gates_data_G[q] = sums[3];
+            gates_data_I[q + 1] = sums[4];
+            gates_data_F[q + 1] = sums[5];
+            gates_data_O[q + 1] = sums[6];
+            gates_data_G[q + 1] = sums[7];
```

Comment:
The default weight memory layout is not well suited for SIMD optimization. Reorder weight layout will probably bring another speedup I guess.
For the current implementation, I think it is still good and working good :smiley:

Reply:
I agree, preferably 8 weights for the  I F O G  would be located in a consecutive memory space. This is possible but I didn't find a clean way of doing it yet, I think it should be possible with some reshapes and transposes.